sources/api/thar-be-updates/src/status.rs

```diff
@@ -0,0 +1,304 @@
+        let active_set = gpt_state.active();
+        let next_set = gpt_state.next().context(error::NoneSetToBoot)?;
+        self.active_partition = Some(StagedImage {
+            image: active_image,
+            next_to_boot: active_set == next_set,
+        });
+        Ok(())
+    }
+
+    /// Sets the staging partition image information
+    pub fn set_staging_partition_image_info(&mut self, image: UpdateImage) {
+        self.staging_partition = Some(StagedImage {
+            image,
+            next_to_boot: false,
+        });
+    }
+
+    /// Mark staging partition as next to boot
+    pub fn mark_staging_partition_next_to_boot(&mut self) -> Result<()> {
+        if let Some(staging_partition) = &mut self.staging_partition {
+            staging_partition.set_next_to_boot(true);
+        } else {
+            return error::StagingPartition {}.fail();
+        }
+        if let Some(active_partition) = &mut self.active_partition {
+            active_partition.set_next_to_boot(false);
+        } else {
+            return error::ActivePartition {}.fail();
+        }
+        Ok(())
+    }
+
+    /// Unmark staging partition as next to boot
+    pub fn unmark_staging_partition_next_to_boot(&mut self) -> Result<()> {
+        if let Some(staging_partition) = &mut self.staging_partition {
+            staging_partition.set_next_to_boot(false);
+        } else {
+            return error::StagingPartition {}.fail();
+        }
+        if let Some(active_partition) = &mut self.active_partition {
+            active_partition.set_next_to_boot(true);
+        } else {
+            return error::ActivePartition {}.fail();
+        }
+        Ok(())
+    }
+
+    /// Sets information regarding the latest command invocation
+    /// Derive success/failure status from exit status when possible.
+    pub fn set_recent_command_info(&mut self, cmd_type: UpdateCommand, cmd_output: &Output) {
+        let exit_status = match cmd_output.status.code() {
+            Some(code) => code,
+            None => cmd_output.status.signal().unwrap_or(1),
+        };
+        let command_result = CommandResult {
+            cmd_type,
+            cmd_status: if exit_status == 0 {
+                CommandStatus::Success
+            } else {
+                CommandStatus::Failed
+            },
+            timestamp: Utc::now(),
+            exit_status: Some(exit_status),
+            stderr: Some(String::from_utf8_lossy(&cmd_output.stderr).to_string()),
+        };
+        self.most_recent_command = Some(command_result);
+    }
+
+    /// Returns the update information of the 'latest' available update
+    pub fn get_latest_update(
+        updates: Vec<update_metadata::Update>,
+    ) -> Result<Option<update_metadata::Update>> {
+        let os_info = BottlerocketRelease::new().context(error::ReleaseVersion)?;
+        for update in updates {
+            // If the current running version is greater than the max version ever published,
```

Comment:
Not sure I follow the reasoning here, we might be thinking about "latest" in different ways.

I'd say that if the current running version is greater than anything in the repo, one of two things might have happened:
* we're running a development build and it's pointing at the production repos
* we're running a production release and we've removed an update from the production repos

In the first case, if I had the operator running on my cluster, I wouldn't want it to drive an automated downgrade. It's possible to work around this by pointing at a different repo, but it would be frustrating.

In the second case, it's less clear. Looking at reasons we might pull an update, it'd be because it's bricking hosts for some reason, or because they upgrade successfully but have some critical regression.

In the "brick" case, for hosts that successfully upgraded we probably wouldn't want them to downgrade automatically - they'd lose any features or security fixes in the upgrade, and they might run the risk of bricking on downgrade.

For the "regression" case, we'd be more likely to fast follow with a new release that reverted the problematic change. There might not be much of a time window where automated downgrade would help, and if the regression only affected a subset of customers, everyone else might still want the upgrade.

So I'd interpret "latest" as `max(os_version, repo_version)` and if the host is running a newer version, then there's no "upgrade" available for them. On the other hand, if we're told to lock to a specific version, we should always take that version (and only that version) as an upgrade, unless we're already running it.

Reply:
This part of the code is trying to mirror `updog`'s behavior for determining what update to update to when calling `updog whats` or `updog update`. 

https://github.com/bottlerocket-os/bottlerocket/blob/30358c86ae797e784ef749237103878ea26f100f/sources/updater/updog/src/main.rs#L160-L168

I should rely on `updog whats` to tell me what it thinks the latest version is instead of trying to duplicate the behavior here.

It sounds like this would need to be fixed in `updog`. Unless `updog`'s perception on what latest is should be different from `version-lock`ed latest?

Reply:
It seems to me like Ben's point should apply the same way to updog, so we'd want a fix there?

Reply:
Had an offline discussion regarding this; `max_version` is needed here for adhering to the Bottlerocket update system's security model.

Keeping the logic here as is for now. Will come back to this once https://github.com/bottlerocket-os/bottlerocket/issues/952 is addressed.